### PR TITLE
SI-9044 Fix order of interfaces in classfiles

### PR DESCRIPTION
**Pull request by lrytz**
It was reversed since ced3ca8. The reason is that the backend used
`mixinClasses` to obtain the parents of a class, which returns them in
linearization order.

`mixinClasses` als returns all ancestors (not only direct parents),
which means more work for `minimizeInterfaces`. This was introduced
in cd62f52 for unclear reasons. So we switch back to using `parents`.

To verify, I created a git repo with the library classfiles of 2.11.4, 2.11.x and this fix (classfile diffing by https://github.com/retronym/libscala/tree/master/git-java)

```
lucmac:t9044 luc$ git lg
* 02a3531 - (HEAD, master) 9044 fix 2 (37 minutes ago) <Lukas Rytz>
* 68b55fe - current 2.11.x (17 hours ago) <Lukas Rytz>
* 405c74d - 2.11.4 (17 hours ago) <Lukas Rytz>
```

the result of diffing 2.11.4 versus this fix (`git diff 405c74d..02a3531 > diff`) is here: https://gist.github.com/lrytz/3fcb0547849cf7b7ef36. The different diff in signatures (`+  Signature: J`) is due to the fix ced3ca8, which minimized the interfaces - the signature diff is not correctly displayed. Here's the diff for List.class (obtained with asm):

``` diff
@@ -1,11 +1,11 @@
 // class version 50.0 (50)
 // access flags 0x421
-// signature <A:Ljava/lang/Object;>Lscala/collection/AbstractSeq<TA;>;Lscala/collection/immutable/LinearSeq<TA;>;Lscala/Product;Lscala/collection/LinearSeqOptimized<TA;Lscala/collection/immutable/List<TA;>;>;Ljava/io/Serializable;
-// declaration: scala/collection/immutable/List<A> extends scala.collection.AbstractSeq<A> implements scala.collection.immutable.LinearSeq<A>, scala.Product, scala.collection.LinearSeqOptimized<A, scala.collection.immutable.List<A>>, java.io.Serializable
+// signature <A:Ljava/lang/Object;>Lscala/collection/AbstractSeq<TA;>;Lscala/collection/immutable/LinearSeq<TA;>;Lscala/Product;Lscala/collection/generic/GenericTraversableTemplate<TA;Lscala/collection/immutable/List;>;Lscala/collection/LinearSeqOptimized<TA;Lscala/collection/immutable/List<TA;>;>;Ljava/io/Serializable;
+// declaration: scala/collection/immutable/List<A> extends scala.collection.AbstractSeq<A> implements scala.collection.immutable.LinearSeq<A>, scala.Product, scala.collection.generic.GenericTraversableTemplate<A, scala.collection.immutable.List>, scala.collection.LinearSeqOptimized<A, scala.collection.immutable.List<A>>, java.io.Serializable
 public abstract class scala/collection/immutable/List extends scala/collection/AbstractSeq  implements scala/collection/immutable/LinearSeq scala/Product scala/collection/LinearSeqOptimized java/io/Serializable  {
```

The gist shows that the list of implemented interfaces is the same for all classes in the library:

```
lucmac:t9044 luc$ cat diff | grep implements | grep +
lucmac:t9044 luc$ 
```

### CODE REVIEW

**Comment by adriaanm** [comment]
Good catch -- how come this never triggered mima?

**Comment by lrytz** [comment]
It seems to [ignore interface order](https://github.com/typesafehub/migration-manager/blob/f0743bf3494aca92e9d355d1ffdbf4378afd7dc5/reporter/src/main/scala/com/typesafe/tools/mima/lib/analyze/template/TemplateRules.scala#L67). Maybe that's fine by the definition of binary compatibility?

``` scala
    def diff(thisTypes: Iterable[ClassInfo], thatTypes: Iterable[ClassInfo]) = {
      val thisSuperclasses = thisTypes.map(_.fullName).toSet

...

      val missing = diff(thisClass.allInterfaces, thatClass.allInterfaces)
```

**Comment by adriaanm** [comment]
Can the order influence whether to emit invokevirtual/invokeinterface?

**Comment by lrytz** [comment]
I don't think so. it's only about the order of interfaces, the super class was always the same.

**Comment by lrytz** [comment]
review by @adriaanm

File: src/compiler/scala/tools/nsc/transform/Erasure.scala [hunk 1]
```diff
@@ -185,22 +185,21 @@ abstract class Erasure extends AddInterfaces
...
-    while(!rest.isEmpty) {
+  def minimizeParents(parents: List[Type]): List[Type] = {
+    var rest   = parents
+    var leaves = collection.mutable.ListBuffer.empty[Type]
+    while(rest.nonEmpty) {
       val candidate = rest.head
       val nonLeaf = leaves exists { t => t.typeSymbol isSubClass candidate.typeSymbol }
       if(!nonLeaf) {
-        leaves = candidate :: (leaves filterNot { t => candidate.typeSymbol isSubClass t.typeSymbol })
+        leaves = (leaves filterNot { t => candidate.typeSymbol isSubClass t.typeSymbol }) += candidate
```
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/transform/Erasure.scala]: Is this assignment to leaves necessary?

Comment (adriaanm) [inline src/compiler/scala/tools/nsc/transform/Erasure.scala]: I wonder whether we can eliminate one of the mutations. `leaves` is a `var` that points to a mutable collection.

Comment (lrytz) [inline src/compiler/scala/tools/nsc/transform/Erasure.scala]: ups :blush:

Comment (adriaanm) [inline src/compiler/scala/tools/nsc/transform/Erasure.scala]: added to the Abide hall of fame :-)

**Comment by lrytz** [comment]
hah!

```
java.lang.IncompatibleClassChangeError: Class scala.collection.mutable.HashMap does not implement the requested interface scala.collection.mutable.HashTable
```

File: src/compiler/scala/tools/nsc/transform/Erasure.scala (hunk 1 above)
Comment (lrytz) [inline src/compiler/scala/tools/nsc/transform/Erasure.scala]: funny enough, the `var` was necessary after all: filterNot returns a new buffer.

**Comment by adriaanm** [comment]
LGTM

**Pull request merged**